### PR TITLE
Added RPCs for the API Key mechanism

### PR DESCRIPTION
**Pull request by dylanratcliffe**
Here's my first stab at the API Key RPCs. Let me know what you think. The callback URL isn't included as the request won't be able to be parsed as a native Connect request, so we'll have to handle that manually. Shouldn't be too hard though

### CODE REVIEW

File: apikey.proto [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    // The API Key that you want to exchange for an Oauth access token
+    string apiKey = 1;
+}
+
+message ExchangeKeyForTokenResponse {
+    // The access token that can now be use to authenticate to Overmind and its
+    // APIs
+    string accessToken = 1;
+}
+
```
Comment (DavidS-om) [inline apikey.proto:124]: Move this to the bottom, to preserve the order of RPCs

```suggestion
```

File: apikey.proto [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+message ListAPIKeysResponse {
+    repeated APIKey keys = 1;
+}
+
+message DeleteAPIKeyRequest {
+    // The UUID of the API key to delete
+    bytes uuid = 1;
+}
+
+message DeleteAPIKeyResponse {}
```
Comment (DavidS-om) [inline apikey.proto:167]:
```suggestion
message DeleteAPIKeyResponse {}

message ExchangeKeyForTokenRequest {
    // The API Key that you want to exchange for an Oauth access token
    string apiKey = 1;
}

message ExchangeKeyForTokenResponse {
    // The access token that can now be use to authenticate to Overmind and its
    // APIs
    string accessToken = 1;
}
```

File: apikey.proto [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    string redirectURL = 3;
+}
+
+message CreateAPIKeyResponse {
+    // Details of the newly created API Key
+    APIKey key = 1;
+
+    // The URL that the user should visit in order to authorize the newly
+    // created key. This will allow Auth0 to generate the access and refresh
+    // tokens and pass them to the API server. The user will be redirected once
```
Comment (DavidS-om) [inline apikey.proto:142]:
```suggestion
    // tokens and pass them to the API server. The user will be redirected back to the frontend once
```

??

File: apikey.proto [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+message CreateAPIKeyRequest {
+    // The name of the key to create
+    string name = 1;
+
+    // The scopes that the key should have
+    repeated string scopes = 2;
+
+    // The URL that the user should be redirected to after authorizing the key
+    string redirectURL = 3;
```
Comment (DavidS-om) [inline apikey.proto:133]:
```suggestion
```

The redirectUrl is a deployment-specific implementation detail of the ApiKey service and doesn't need to (can't!) be supplied by an external client, therefore it doesn't need to be supplied here.

**Review by DavidS-om — CHANGES_REQUESTED** [review]
Besides a few editorial changes, one big one re the redirect URL: Since that has to be configured in the Auth0 application config, we can't specify it here. It'll have to be something hardcoded (from the deployment config) where we correlate the incoming request through the attached `state`. Then we can forward the user back to the frontend.

Writing this I realise that that was maybe the intention, but that doesn't come through in the comments and re-using the OAuth terminology doesn't help either. In that case it should be something more specific like `frontendLandingPage`.

**Comment by dylanratcliffe** [comment]
You're absolutely right. Good catch with the `state` param. I've updated the comments to reflect the static URL end use of the `state` param to determine the UUID, and also renamed the `redirectURL` to `finalFrontendRedirect` to distinguish it from OAuth stuff

**Review by DavidS-om — APPROVED** [review]

**Pull request merged**